### PR TITLE
[Text Analytics] Add support for healthcare features

### PR DESCRIPTION
**Pull request by joseharriaga**
Resolves:
- [x] #31269 
- [x] #31274
- [x] #31275
- [x] #28811

Features added:
- Added `FhirVersion` and `DocumentType` properties to `AnalyzeHealthcareEntitiesOptions` and `AnalyzeHealthcareEntitiesAction`.
- Added `FhirBundle` property to `AnalyzeHealthcareEntitiesResult`.
- Added `ConfidenceScore` property to `HealthcareEntityRelation`.
- Added `WellKnownFhirVersion` and `HealthcareDocumentType` enums.

### CODE REVIEW

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/Transforms.Legacy.cs [hunk 1]
```diff
@@ -704,8 +705,9 @@ private static IList<HealthcareEntityRelation> ConvertToHealthcareEntityRelation
             foreach (var relation in healthcareRelations)
             {
                 result.Add(new HealthcareEntityRelation(
-                                    relation.RelationType.ToString(),
-                                    ConvertToHealthcareEntityRelationRoleCollection(relation.Entities, healthcareEntities)));
+                    relation.RelationType.ToString(),
+                    ConvertToHealthcareEntityRelationRoleCollection(relation.Entities, healthcareEntities),
+                    default));
```
[the remark this answers is not in the document]
Comment (heaths) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/Transforms.Legacy.cs:710]: Note: our [Azure SDK for .NET guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html) to state they pull from .NET guidelines which used to be https://learn.microsoft.com/dotnet/standard/design-guidelines/ (specifically, the non-null collection properties comes from https://learn.microsoft.com/dotnet/standard/design-guidelines/guidelines-for-collections#collection-properties-and-return-values). However, these have been superseded by new guidelines available from O'Reilly as linked in our guidelines at https://aka.ms/fxdg3. Feel free to ask in our internal Teams team about how to get access.

**Review by heaths — APPROVED** [review]

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs [hunk 2]
```diff
@@ -1287,6 +1321,23 @@ public enum TextSentiment
         Negative = 2,
         Mixed = 3,
     }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct WellKnownFhirVersion : System.IEquatable<Azure.AI.TextAnalytics.WellKnownFhirVersion>
+    {
+        private readonly object _dummy;
+        private readonly int _dummyPrimitive;
+        public WellKnownFhirVersion(string value) { throw null; }
```
Comment (mssfang) [inline sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:1329]: Curious question: When .net use wellKnownXXX for the naming

**Review by kristapratico — APPROVED** [review]

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs (hunk 2 above)
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:1329]: I got the name from the thread in this open issue: https://github.com/Azure/azure-sdk-for-net/issues/28811

Comment (deyaaeldeen) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:1329]: Yeah it is a .NET-specific pattern according to @heaths, see https://github.com/Azure/azure-sdk-for-net/issues/28811#issuecomment-1157098236. In JS, the corresponding pattern is `Known<name>`.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsModelFactory.cs [hunk 3]
```diff
@@ -1319,7 +1319,28 @@ internal static LinkedEntityMatch LinkedEntityMatch(double confidenceScore = def
...
+        /// Initializes a new instance of <see cref="TextAnalytics.AnalyzeHealthcareEntitiesResult"/> for mocking purposes.
+        /// </summary>
+        /// <param name="id">Sets the <see cref="TextAnalyticsResult.Id"/> property.</param>
+        /// <param name="statistics">Sets the <see cref="TextAnalyticsResult.Statistics"/> property.</param>
+        /// <param name="healthcareEntities">Sets the collection of <see cref="TextAnalytics.HealthcareEntity"/>.</param>
+        /// <param name="entityRelations">Sets the collection of <see cref="TextAnalytics.HealthcareEntityRelation"/>.</param>
+        /// <param name="warnings">Sets the collection of <see cref="TextAnalytics.TextAnalyticsWarning"/>.</param>
+        /// <param name="fhirBundle">Sets the <see cref="AnalyzeHealthcareEntitiesResult.FhirBundle"/> property.</param>
+        /// <returns>A new instance of <see cref="TextAnalytics.AnalyzeHealthcareEntitiesResult"/> for mocking purposes.</returns>
+        public static AnalyzeHealthcareEntitiesResult AnalyzeHealthcareEntitiesResult(
```
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsModelFactory.cs:1335]: Thank you for the links! I created the following issue and added it to the "Backlog" milestone: https://github.com/Azure/azure-sdk-for-net/issues/31945

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs [hunk 4]
```diff
@@ -87,12 +89,15 @@ public partial class AnalyzeHealthcareEntitiesOptions : Azure.AI.TextAnalytics.T
...
         public string DisplayName { get { throw null; } set { } }
+        public Azure.AI.TextAnalytics.HealthcareDocumentType? DocumentType { get { throw null; } set { } }
+        public Azure.AI.TextAnalytics.WellKnownFhirVersion? FhirVersion { get { throw null; } set { } }
     }
     public partial class AnalyzeHealthcareEntitiesResult : Azure.AI.TextAnalytics.TextAnalyticsResult
     {
         internal AnalyzeHealthcareEntitiesResult() { }
         public System.Collections.Generic.IReadOnlyCollection<Azure.AI.TextAnalytics.HealthcareEntity> Entities { get { throw null; } }
         public System.Collections.Generic.IReadOnlyCollection<Azure.AI.TextAnalytics.HealthcareEntityRelation> EntityRelations { get { throw null; } }
+        public System.Collections.Generic.IReadOnlyDictionary<string, object> FhirBundle { get { throw null; } }
```
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:100]: I agree that there are probably better ways of exposing this property. For the purposes of this preview of the client library, I'm thinking it's okay to keep like this, but I have opened an issue to revisit this later: https://github.com/Azure/azure-sdk-for-net/issues/31946

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/AnalyzeHealthcareEntitiesResult.cs [hunk 5]
```diff
@@ -18,15 +18,19 @@ public partial class AnalyzeHealthcareEntitiesResult : TextAnalyticsResult
...
         /// </summary>
-        internal AnalyzeHealthcareEntitiesResult(string id, TextDocumentStatistics statistics,
+        internal AnalyzeHealthcareEntitiesResult(
+            string id,
+            TextDocumentStatistics statistics,
             IList<HealthcareEntity> healthcareEntities,
             IList<HealthcareEntityRelation> entityRelations,
-            IList<TextAnalyticsWarning> warnings)
+            IList<TextAnalyticsWarning> warnings,
+            IDictionary<string, object> fhirBundle)
```
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/AnalyzeHealthcareEntitiesResult.cs:27]: I added a check for null to create an empty collection.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/Transforms.Legacy.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/Transforms.Legacy.cs:710]: Thanks for catching this one. I changed the constructor to check for null and create an empty dictionary. I also modified the tests to exercise both, the new transform but also this legacy transform.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/AnalyzeHealthcareEntitiesTests.cs [hunk 6]
```diff
@@ -133,6 +133,10 @@ public async Task AnalyzeHealthcareEntitiesTest()
                     Assert.AreEqual(5, role.Entity.Length);
                 }
             }
+
+            // Check that the FHIR bundle is not null, but empty.
+            Assert.IsNotNull(result1.FhirBundle);
+            Assert.AreEqual(result1.FhirBundle.Count, 0);
```
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/AnalyzeHealthcareEntitiesTests.cs:139]: Nit: with NUnit, the first argument is the expected value (changes how the error message is when the assertion fails) unless you use `Assert.That`.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/AnalyzeHealthcareEntitiesResult.cs [hunk 7]
```diff
@@ -30,7 +30,9 @@ public partial class AnalyzeHealthcareEntitiesResult : TextAnalyticsResult
             _entities = new ReadOnlyCollection<HealthcareEntity>(healthcareEntities);
             Warnings = new ReadOnlyCollection<TextAnalyticsWarning>(warnings);
             EntityRelations = new ReadOnlyCollection<HealthcareEntityRelation>(entityRelations);
-            FhirBundle = new ReadOnlyDictionary<string, object>(fhirBundle);
+            FhirBundle = (fhirBundle is not null)
```
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/AnalyzeHealthcareEntitiesResult.cs:33]: TDD for the win, eh? :smile:

**Review by heaths — APPROVED** [review]

**Review by heaths — APPROVED** [review]

**Pull request merged**